net/fetcher.go

```diff
@@ -22,6 +30,212 @@ type Fetcher interface {
 	FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(ts types.TipSet) (bool, error)) ([]types.TipSet, error)
 }
 
+// GraphSyncFetcher is used to fetch data over the network.  It is implemented with
+// a persistent bitswap session on a networked blockservice.
+type GraphSyncFetcher struct {
+	gs        *graphsync.GraphSync
+	validator consensus.BlockSyntaxValidator
+	bridge    ipldbridge.IPLDBridge
+	bs        bstore.Blockstore
+}
+
+// NewGraphSyncFetcher returns a BitswapFetcher wired up to the input BlockService and a newly
+// initialized persistent session of the block service.
+func NewGraphSyncFetcher(ctx context.Context, network gsnet.GraphSyncNetwork, bridge ipldbridge.IPLDBridge, blockstore bstore.Blockstore,
+	bv consensus.BlockSyntaxValidator) *GraphSyncFetcher {
+	gsf := &GraphSyncFetcher{
+		bs:        blockstore,
+		validator: bv,
+		bridge:    bridge,
+	}
+	gsf.gs = graphsync.New(ctx, network, bridge, gsf.loader, gsf.storer)
+	return gsf
+}
+
+func (gsf *GraphSyncFetcher) loader(lnk ipldp.Link, lnkCtx ipldp.LinkContext) (io.Reader, error) {
+	asCidLink, ok := lnk.(cidlink.Link)
+	if !ok {
+		return nil, fmt.Errorf("Unsupported Link Type")
+	}
+	block, err := gsf.bs.Get(asCidLink.Cid)
+	if err != nil {
+		return nil, err
+	}
+	return bytes.NewReader(block.RawData()), nil
+}
+
+func (gsf *GraphSyncFetcher) storer(lnkCtx ipldp.LinkContext) (io.Writer, ipldp.StoreCommitter, error) {
+	var buffer bytes.Buffer
+	committer := func(lnk ipldp.Link) error {
+		asCidLink, ok := lnk.(cidlink.Link)
+		if !ok {
+			return fmt.Errorf("Unsupported Link Type")
+		}
+		block, err := blocks.NewBlockWithCid(buffer.Bytes(), asCidLink.Cid)
+		if err != nil {
+			return err
+		}
+		return gsf.bs.Put(block)
+	}
+	return &buffer, committer, nil
+}
+
+const maxRecursionDepth = 64
+const recursionMultiplier = 4
+
+// FetchTipSets gets Tipsets starting from the given tipset key and continuing till
+// the done function returns true or errors
+func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	err := gsf.GetSingleLayer(ctx, cids, from)
```

Comment:
I don't understand why we are doing a single request here. I initially thought we needed the CID's of `tsKey` parents, but that doesn't seem to be the case since we are passing `cids` to `GetBlocks` below. Can you add a comment explaining what is going on here?

Reply:
Per our last synchronous chat, this is required since graphsync operates over a single CID. No clarification/reply here needed, but a comment stating this would be appreciated.